File: metal/paddle-mobile/paddle-mobile/Src/Operators/Kernels/LeakyReluKernel.swift
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */

import Foundation

struct LeakyReluMetalParam {
    let alpha: Float32
}

class LeakyReluKernel<P: PrecisionProtocol>: Kernel, Computable {
    var metalParam: LeakyReluMetalParam
    func compute(commandBuffer: MTLCommandBuffer, param: LeakyReluParam<P>) throws {
        guard let tempPipline = pipline else {
            throw PaddleMobileError.makeError(type: .predictError, msg: "pipline is nil")
        }
        guard let inputTexture = param.input.metalTexture else {
            throw PaddleMobileError.makeError(type: .predictError, msg: "input metaltexture is nil")
        }
        guard let outputTexture = param.output.metalTexture else {
            throw PaddleMobileError.makeError(type: .predictError, msg: "output metaltexture is nil")
        }
        do {
            guard let encoder = commandBuffer.makeComputeCommandEncoder() else {
                throw PaddleMobileError.makeError(type: .predictError, msg: "encoder is nil")
            }
            defer {
                encoder.endEncoding()
            }
            encoder.setTexture(inputTexture, index: 0)
            encoder.setTexture(outputTexture, index: 1)
            encoder.setBytes(&metalParam, length: MemoryLayout<Relu6MetalParam>.size, index: 0)
            try encoder.dispatch(computePipline: tempPipline, outTexture: outputTexture)
        }
    }
    
    required init(device: MTLDevice, param: LeakyReluParam<P>, initContext: InitContext) throws {
        try param.output.initTexture(device: device, inTranspose: param.input.transpose, computePrecision: GlobalConfig.shared.computePrecision)
        metalParam = LeakyReluMetalParam(alpha: param.alpha)
        if GlobalConfig.shared.computePrecision == .Float32 {
            try super.init(device: device, inFunctionName: "leaky_relu", initContext: initContext)
        } else if GlobalConfig.shared.computePrecision == .Float16 {
            try super.init(device: device, inFunctionName: "leaky_relu_half", initContext: initContext)
        } else {
            throw PaddleMobileError.makeError(type: .predictError, msg: "unsupported compute precision: \(GlobalConfig.shared.computePrecision)")
        }
    }
}

File: src/operators/kernel/cl/leakyrelu_kernel.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#ifdef LEAKY_RELU_OP

#include <operators/kernel/activation_kernel.h>

namespace paddle_mobile {
namespace operators {
template <>
bool LeakyReluKernel<GPU_CL, float>::Init(
    paddle_mobile::operators::LeakyReluParam<paddle_mobile::GPU_CL> *param) {
  this->cl_helper_.AddKernel("leakyrelu", "leakyrelu_kernel.cl");
  return true;
}

template <>
void LeakyReluKernel<GPU_CL, float>::Compute(
    const paddle_mobile::operators::LeakyReluParam<paddle_mobile::GPU_CL>
        &param) {
  auto kernel = this->cl_helper_.KernelAt(0);
  auto default_work_size = this->cl_helper_.DefaultWorkSize(*(param.Out()));
  auto input = param.InputX();
  cl_mem input_image = input->GetCLImage();
  auto output = param.Out();
  cl_mem out_image = output->GetCLImage();
  float alpha = param.Alpha();
  int out_dims_w = output->dims()[3];

  cl_int status;
  status = clSetKernelArg(kernel, 0, sizeof(cl_mem), &input_image);
  CL_CHECK_ERRORS(status);
  status = clSetKernelArg(kernel, 1, sizeof(cl_mem), &out_image);
  CL_CHECK_ERRORS(status);
  status = clSetKernelArg(kernel, 2, sizeof(float), &alpha);
  CL_CHECK_ERRORS(status);
  status = clSetKernelArg(kernel, 3, sizeof(int), &out_dims_w);
  CL_CHECK_ERRORS(status);
  status = clEnqueueNDRangeKernel(
      this->cl_helper_.CLCommandQueue(), kernel, default_work_size.size(), NULL,
      default_work_size.data(), NULL, 0, NULL, NULL);
  CL_CHECK_ERRORS(status);
}
template class LeakyReluKernel<GPU_CL, float>;
}  // namespace operators
}  // namespace paddle_mobile

#endif
